src/main/scala/org/ergoplatform/nodeView/history/extra/IndexedErgoAddress.scala

```diff
@@ -42,24 +47,84 @@ case class IndexedErgoAddress(treeHash: ModifierId,
     */
   def boxCount(): Long = segmentTreshold * boxSegmentCount + boxes.length
 
+  /**
+    * Copied from [[java.util.Arrays.binarySearch]]
+    */
+  private def binarySearch(a: ArrayBuffer[Long], key: Long): Int = {
+    var low: Int = 0
+    var high: Int = a.length - 1
+
+    while (low <= high) {
+      val mid = (low + high) >>> 1
+      val midVal = abs(a(mid)) // ignore negativity
+
+      if (midVal < key)
+        low = mid + 1
+      else if (midVal > key)
+        high = mid - 1
+      else
+        return mid // key found
+    }
+    -1 // key not found.
+  }
+
+  /**
+    * Retrieve segment with specified id from buffer or database
+    * @param history - database handle to search, if segment is not found in buffer
+    * @param id - address segment to search for
+    * @return
+    */
+  private def getSegmentFromBufferOrHistroy(history: ErgoHistoryReader, id: ModifierId): Int = {
+    cfor(segments.length - 1)(_ >= 0, _ - 1) { i =>
+      if(segments(i).id.equals(id)) return i
+    }
+    segments += history.typedExtraIndexById[IndexedErgoAddress](id).get
+    segments.length - 1
+  }
+
+  /**
+    * Locate which segment the given box number is in and change its sign, meaning it spends unspent boxes and vice versa.
+    * @param boxNum - box number to locate
+    * @param history - database to retrieve segments from
+    */
+  private[extra] def findAndModBox(boxNum: Long, history: ErgoHistoryReader): Unit = {
+    val boxNumAbs = abs(boxNum)
+    val inCurrent: Int = binarySearch(boxes, boxNumAbs)
+    if(inCurrent >= 0) { // box found in current box array
+      boxes(inCurrent) = -boxes(inCurrent)
+    } else { // box is in another segment, use binary search to locate
+      var n = 0
+      var low = 0
+      var high = boxSegmentCount - 1
+      while(low <= high) {
+        val mid = (low + high) >>> 1
+        n = getSegmentFromBufferOrHistroy(history, boxSegmentId(treeHash, mid))
+        if(abs(segments(n).boxes.head) < boxNumAbs &&
+           abs(segments(n).boxes.last) < boxNumAbs)
+          low = mid + 1
+        else if(abs(segments(n).boxes.head) > boxNumAbs &&
+                abs(segments(n).boxes.last) > boxNumAbs)
+          high = mid - 1
+        else
+          low = high + 1 // break
+      }
+      val i: Int = binarySearch(segments(n).boxes, boxNumAbs)
+      if(i >= 0)
+        segments(n).boxes(i) = -segments(n).boxes(i)
+      else
+        log.warn(s"Box $boxNum not found in any segment of parent address when trying to spend")
+    }
+  }
```

Comment:
This is quite complex piece of logic. Better to have unit tests for such methods.

Reply:
Now covered in tests

Reply:
https://github.com/ergoplatform/ergo/pull/1968/files#diff-35d4f2ccbf422c14ab6c7438701f05b00aaba8087e1df0f1a72de6b432412c05R100
If the logic fails to modify the box ids this will fail

Reply:
But findAndModBox is not called there.

Reply:
Its called in https://github.com/ergoplatform/ergo/pull/1968/files#diff-35d4f2ccbf422c14ab6c7438701f05b00aaba8087e1df0f1a72de6b432412c05R174